dm/master/server.go

```diff
@@ -1530,73 +1533,99 @@ func (s *Server) waitOperationOk(ctx context.Context, cli *scheduler.Worker, tas
 					SourceStatus: &pb.SourceStatus{Source: sourceID, Worker: cli.BaseInfo().Name},
 				}
 				if w := s.scheduler.GetWorkerByName(cli.BaseInfo().Name); w == nil {
-					return resp, nil
+					return true, "", resp, nil
 				} else if cli.Stage() == scheduler.WorkerOffline {
-					return resp, nil
+					return true, "", resp, nil
 				}
 			}
 		}
-		var queryResp *pb.QueryStatusResponse
+
 		resp, err := cli.SendRequest(ctx, req, s.cfg.RPCTimeout)
 		if err != nil {
-			log.L().Error("fail to query operation", zap.Int("retryNum", num), zap.String("task", taskName),
-				zap.String("source", sourceID), zap.Stringer("expect", expect), log.ShortError(err))
+			log.L().Error("fail to query operation",
+				zap.Int("retryNum", num),
+				zap.String("task", taskName),
+				zap.String("source", sourceID),
+				zap.Stringer("expect", expect),
+				log.ShortError(err))
 		} else {
-			queryResp = resp.QueryStatus
+			queryResp := resp.QueryStatus
+			if queryResp == nil {
+				// should not happen
+				return false, "", nil, errors.Errorf("expect a query-status response, got type %v", resp.Type)
+			}
 
 			switch masterReq.(type) {
 			case *pb.OperateSourceRequest:
+				if queryResp.SourceStatus == nil {
+					continue
+				}
 				switch expect {
 				case pb.Stage_Running:
 					if queryResp.SourceStatus.Source == sourceID {
-						if err := extractWorkerError(queryResp.SourceStatus.Result); err != nil {
-							return queryResp, err
+						msg := ""
+						if err2 := extractWorkerError(queryResp.SourceStatus.Result); err2 != nil {
+							msg = err2.Error()
 						}
-						return queryResp, nil
+						return true, msg, queryResp, nil
 					}
 				case pb.Stage_Stopped:
 					// we don't use queryResp.SourceStatus.Source == "" because worker might be re-arranged after being stopped
 					if queryResp.SourceStatus.Source != sourceID {
-						if err := extractWorkerError(queryResp.SourceStatus.Result); err != nil {
-							return queryResp, err
+						msg := ""
+						if err2 := extractWorkerError(queryResp.SourceStatus.Result); err2 != nil {
+							msg = err2.Error()
 						}
-						return queryResp, nil
+						return true, msg, queryResp, nil
 					}
 				}
 			case *pb.StartTaskRequest, *pb.UpdateTaskRequest, *pb.OperateTaskRequest:
 				if expect == pb.Stage_Stopped && len(queryResp.SubTaskStatus) == 0 {
-					return queryResp, nil
-				} else if len(queryResp.SubTaskStatus) == 1 {
+					return true, "", queryResp, nil
+				}
+				if len(queryResp.SubTaskStatus) == 1 {
 					if subtaskStatus := queryResp.SubTaskStatus[0]; subtaskStatus != nil {
-						if err := extractWorkerError(subtaskStatus.Result); err != nil {
-							return queryResp, err
+						msg := ""
+						if err2 := extractWorkerError(subtaskStatus.Result); err2 != nil {
+							msg = err2.Error()
 						}
+						ok := false
```

Comment:
maybe use `ok := msg == ""`, so that we don't need to check msg!="" before return

Reply:
the two variables are independent

Reply:
i got it, there is a situation that: ok=false but msg contains some message need to return